File: aoc/aoc24/d4/part1.rs
const DELTAS: [(isize, isize); 8] = [(1, 0), (-1, 0), (0, 1), (0, -1), (1, 1), (1, -1), (-1, 1), (-1, -1)];
const PATTERN: [char; 4] = ['X', 'M', 'A', 'S'];

fn main() {
    let mut grid: Vec<Vec<char>> = Vec::new();
    for line in std::io::stdin().lines().map(|x| x.unwrap()) {
        grid.push(line.chars().collect());
    }

    let mut total = 0;

    let rows = grid.len() as isize;
    let cols = grid[0].len() as isize;
    for i in 0..rows {
        for j in 0..cols {
            for (di, dj) in DELTAS {
                let mut count = 0;
                let mut x = i;
                let mut y = j;
                while count < 4 && 0 <= x && x < rows && 0 <= y && y < cols {

                    if PATTERN[count] != grid[x as usize][y as usize] {
                        break
                    }
                    count += 1;
                    x += di;
                    y += dj;
                }
                if count == 4 {
                    total += 1;
                }
            }
        }
    }
    println!("{}", total);
}

File: aoc/aoc25/d4/part1.py
import sys

def valid_moves(i, j):
    l = []
    for x, y in (
        (i + 1, j),
        (i - 1, j),
        (i, j + 1),
        (i, j - 1),
        (i + 1, j + 1),
        (i - 1, j + 1),
        (i + 1, j - 1),
        (i - 1, j - 1),
    ):
        if 0 <= x < rows and 0 <= y < cols:
            l.append((x, y))
    return l

grid = [[c for c in line.rstrip()] for line in sys.stdin]
rows = len(grid)
cols = len(grid[0])

s = 0
for i in range(rows):
    for j in range(cols):
        if grid[i][j] == "@":
            if sum(grid[x][y] == "@" for x, y in valid_moves(i, j)) < 4:
                s += 1
print(s)